ethash/src/compute.rs

```diff
@@ -269,57 +271,81 @@ pub fn light_compute(light: &Light, header_hash: &H256, nonce: u64) -> ProofOfWo
 }
 
 fn hash_compute(light: &Light, full_size: usize, header_hash: &H256, nonce: u64) -> ProofOfWork {
+	macro_rules! make_const_array {
+		($n:expr, $value:expr) => {{
+			// We use explicit lifetimes to ensure that val's borrow is invalidated until the
+			// transmuted val dies.
+			unsafe fn make_const_array<'a, T, U>(val: &'a mut [T]) -> &'a mut [U; $n] {
+				use ::std::mem;
+
+				debug_assert_eq!(val.len() * mem::size_of::<T>(), $n * mem::size_of::<U>());
+				mem::transmute(val.as_mut_ptr())
+			}
+
+			make_const_array($value)
+		}}
+	}
+
 	if full_size % MIX_WORDS != 0 {
 		panic!("Unaligned full size");
 	}
-	// pack hash and nonce together into first 40 bytes of s_mix
-	let mut s_mix: [Node; MIX_NODES + 1] = [Node::default(), Node::default(), Node::default()];
-	unsafe { ptr::copy_nonoverlapping(header_hash.as_ptr(), s_mix.get_unchecked_mut(0).bytes.as_mut_ptr(), 32) };
-	unsafe { ptr::copy_nonoverlapping(mem::transmute(&nonce), s_mix.get_unchecked_mut(0).bytes[32..].as_mut_ptr(), 8) };
 
-	// compute sha3-512 hash and replicate across mix
 	unsafe {
+		// pack hash and nonce together into first 40 bytes of s_mix
+		let mut s_mix: [Node; MIX_NODES + 1] = mem::uninitialized();
+
+		ptr::copy_nonoverlapping(
+			header_hash.as_ptr(),
+			s_mix.get_unchecked_mut(0).bytes.as_mut_ptr(),
+			32,
+		);
+		ptr::copy_nonoverlapping(
+			mem::transmute(&nonce),
+			s_mix.get_unchecked_mut(0).bytes[32..].as_mut_ptr(),
+			8
+		);
```

Comment:
copy_from_slice is exactly the same thing as copy_nonoverlapping [link](https://doc.rust-lang.org/src/core/slice/mod.rs.html#667)